src/lib/cpd/CpdEngine.ts

```diff
@@ -0,0 +1,246 @@
+
+	// eslint-disable-next-line @typescript-eslint/no-explicit-any
+	private jsonToRuleResults(duplications: any): RuleResult[] {    
+		const ruleResults: RuleResult[] = [];
+
+		for (const duplication of duplications) {
+		
+			const parts = duplication.elements;
+			const codeFragment = parts.find(part => {return part.name === 'codefragment'}).elements[0].cdata;
+			const occurences = parts.filter(part => {return part.name === 'file'});
+			const codeFragmentID = checksum(codeFragment).slice(0,7);
+			// TODO: check for collisions (cpd can return multiple duplications for the same code fragment, so these have the same checksum currently)
+
+			let occCount = 1;
+			for (const occ of occurences) {
+				// create a violation for each occurence of the code fragment
+				const violation: RuleViolation = {
+					line: occ.attributes.line,
+					column: occ.attributes.column,
+					endLine: occ.attributes.endline,
+					endColumn: occ.attributes.endcolumn,
+					ruleName: this.cpdCatalog.rules[0].name,
+					severity: 3,
+					message: messages.getMessage("CpdViolationMessage", [codeFragmentID, occCount, occurences.length, duplication.attributes.lines, duplication.attributes.tokens]),
+					category: this.cpdCatalog.categories[0].name,
+					url: 'https://pmd.github.io/latest/pmd_userdocs_cpd.html#refactoring-duplicates',
+				};
+				occCount++;
+
+				// if there are already violations for a file, add to its violations, if not, create a new RuleResult for that file.
+				const ruleResult: RuleResult = ruleResults.find(ruleResult => {return occ.attributes.path === ruleResult.fileName})
+				if (ruleResult) {
+					ruleResult.violations.push(violation);
+				} else {
+					ruleResults.push(
+						{
+							engine: CpdEngine.ENGINE_NAME,
+							fileName: occ.attributes.path,
+							violations: [violation]
+						}
+					);
+				}
+			}
+		}
+		return ruleResults;
+	}
+
+	
+	public async init(): Promise<void> {
+		if (this.initialized) {
+			return;
+		}
+		this.logger = await Logger.child(this.getName());
+		this.config = await Controller.getConfig();
+		this.initialized = true;
+		this.logger = await Logger.child(this.getName())
+
+		const languages: Set<string> = new Set();
+		for (const pattern of await this.config.getTargetPatterns(ENGINE.CPD)){
+			const ext = pattern.substr(pattern.lastIndexOf("."));
+			if (FILE_EXTS_TO_LANGUAGE.has(ext)) {
+				languages.add(FILE_EXTS_TO_LANGUAGE.get(ext));
+			}
+		}
+
+		this.cpdCatalog = {
+			rules: [{
+				engine: ENGINE.CPD.valueOf(),
+				sourcepackage: ENGINE.CPD.valueOf(),
+				name: 'copy-paste-detected',
+				description: 'Identify duplicate code blocks.',
+				categories: ['Copy/Paste Detected'],
+				rulesets: [],
+				languages: Array.from(languages),
+				defaultEnabled: true
+			}],
+			categories: [{
+				engine: ENGINE.CPD.valueOf(),
+				name: 'Copy/Paste Detected',
+				paths: []
+			}],
+			rulesets: []
+		};
+
+		CpdEngine.minimumTokens = await this.config.getMinimumTokens(ENGINE.CPD)
```

Comment:
I think this is missing a terminating semicolon.

@rmohan20 should we consider supporting an environment variable so that users don't have to edit the config? One issue with editing configs is that they don't existing on a brand new CI/CD run.

something like `CpdEngine.minimumTokens = process.env.SFDX_SCANNER_CPD_MIN_TOKENS || (await this.config.getMinimumTokens(ENGINE.CPD));`

Reply:
@jbartolotta-sfdc Agreed. Not for this story, but sounds like a good direction for all our config.